cors/cors.go

```diff
@@ -0,0 +1,174 @@
+		}
+
+		// Headers validation
+		for _, header := range rule.AllowedHeader {
+			if strings.Count(header, "*") > 1 {
+				return fmt.Errorf("header %s in CORS rule, id: %s, %w", header, rule.ID, ErrAllowedHeaderWildcards{Header: header})
+			}
+		}
+	}
+
+	return nil
+}
+
+// HasAllowedOrigin returns true if the given origin is allowed by the CORS rule
+func (c *Rule) HasAllowedOrigin(origin string) bool {
+	// See "AllowedOrigin element" in https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html
+	for _, allowedOrigin := range c.AllowedOrigin {
+		if wildcard.Match(allowedOrigin, origin) {
+			// Only one wildcard character (*) is allowed by S3 spec, but Match does
+			// not enforce that, it's done by Validate() function.
+			// Origins are case sensitive
+			return true
+		}
+	}
+	return false
+}
+
+// HasAllowedMethod returns true if the given method is contained in the CORS rule.
+func (c *Rule) HasAllowedMethod(method string) bool {
+	// See "AllowedMethod element" in https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html
+	for _, allowedMethod := range c.AllowedMethod {
+		if allowedMethod == method {
+			// Methods are always uppercase, enforced by Validate() function.
+			return true
+		}
+	}
+	return false
+}
+
+// FilterAllowedHeaders returns the headers that are allowed by the rule, and a boolean indicating if all headers are allowed.
+func (c *Rule) FilterAllowedHeaders(headers []string) ([]string, bool) {
+	// See "AllowedHeader element" in https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html
+	// It's inefficient to store the CORS config verbatim and run ToLower here, but S3 essentially
+	// behaves this way, and will return the XML config verbatim when you GET it.
+	filtered := []string{}
+	for _, header := range headers {
+		header = strings.ToLower(header)
+		found := false
+		for _, allowedHeader := range c.AllowedHeader {
+			// Case insensitive comparison for headers
+			if wildcard.Match(strings.ToLower(allowedHeader), header) {
+				// Only one wildcard character (*) is allowed by S3 spec, but Match does
+				// not enforce that, it's done by rule.Validate() function.
+				filtered = append(filtered, header)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, false
+		}
+	}
+	return filtered, true
+}
+
+// ParseBucketCorsConfig parses a CORS configuration in XML from an io.Reader.
+func ParseBucketCorsConfig(reader io.Reader) (*Config, error) {
+	var c Config
+	err := xml.NewDecoder(reader).Decode(&c)
+	if err != nil {
+		return nil, fmt.Errorf("decoding xml: %w", err)
+	}
+	if c.XMLNS == "" {
+		c.XMLNS = defaultXMLNS
+	}
+	for i, rule := range c.CORSRules {
+		for j, method := range rule.AllowedMethod {
+			c.CORSRules[i].AllowedMethod[j] = strings.ToUpper(method)
+		}
+	}
+	return &c, nil
+}
+
+// ToXML marshals the CORS configuration to XML.
+func (c *Config) ToXML() ([]byte, error) {
+	if c.XMLNS == "" {
+		c.XMLNS = defaultXMLNS
+	}
```

Comment:
It seems a bit strange that ToXML is mutating the value. Maybe change the signature to `func (c Config) ToXML()`, so you aren't actually mutating the caller?